src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs

```diff
@@ -333,11 +353,121 @@ private async Task<SearchResult> SearchImpl(SearchOptions searchOptions, SqlSear
                                 Core.Resources.TruncatedIncludeMessage));
                     }
 
-                    return new SearchResult(resources, continuationToken?.ToJson(), searchOptions.Sort, searchOptions.UnsupportedSearchParams);
+                    return new SearchResult(resources, continuationToken?.ToJson(), originalSearchOptions.Sort, searchOptions.UnsupportedSearchParams);
                 }
             }
         }
 
+        /// <summary>
+        /// If no sorting fields are specified, sets the sorting fields to the primary key. (This is either ResourceSurrogateId or ResourceTypeId, ResourceSurrogateId).
+        /// If sorting only by ResourceTypeId, adds in ResourceSurrogateId as the second sort column.
+        /// If sorting by ResourceSurrogateId and using partitioned tables and searching over a single type, sets the sort to ResourceTypeId, ResourceSurrogateId
+        /// </summary>
+        /// <param name="searchOptions">The input SearchOptions</param>
+        /// <param name="searchExpression">The searchExpression</param>
+        /// <param name="sqlSearchType">The type of search being performed</param>
+        /// <returns>If the sort needs to be updated, a new <see cref="SearchOptions"/> instance, otherwise, the same instance as <paramref name="searchOptions"/></returns>
+        private SearchOptions UpdateSort(SearchOptions searchOptions, Expression searchExpression, SqlSearchType sqlSearchType)
+        {
+            if (sqlSearchType == SqlSearchType.History)
+            {
+                // history is always sorted by _lastUpdated.
+                searchOptions = searchOptions.Clone();
+
+                ISearchParameterDefinitionManager searchParameterDefinitionManager = _searchParameterDefinitionManagerResolver.Invoke();
+
+                searchOptions.Sort = new (SearchParameterInfo searchParameterInfo, SortOrder sortOrder)[]
+                {
+                    (searchParameterDefinitionManager.GetSearchParameter(KnownResourceTypes.Resource, SearchParameterNames.LastUpdated), SortOrder.Ascending),
```

Comment:
>searchParameterDefinitionManager.GetSearchParameter(KnownResourceTypes.Resource, SearchParameterNames.LastUpdated [](http://example.com/codeflow?start=21&length=113)

is it worth to resolve this and `ResourceType` in constuctor as property and reuse it?

Reply:
It would be nice, but this is currently not registered as a singleton. (but it should be)